### ISSUE TITLE
"Cannot call -(void)open on SRWebSocket more than once"

### ISSUE THREAD

**Issue by PetraKumaSun** [square/SocketRocket#305]
My code

if (myWebSocket.readyState == SR_CLOSED) {
     myWebSocket.delegate = self;
     [myWebSocket open]; 
}

I load the screen with this code. Without waiting for connections, I move to another screen, and then go back again. Checks condition - socket closed! But the application crash with the message 
"NSAssert(_readyState == SR_CONNECTING, @"Cannot call -(void)open on SRWebSocket more than once");" 

What could be the problem?

**Comment by kenwigginton** [comment]
@PetraKumaSun
It doesn't appear that this can actually happen since the `readyState` has to be SR_CONNECTING, for that assertion to occur right? I'm not exactly sure what's happening here, maybe you're calling open from your other screen as well? Are you sure that your code here is actually being executed? Just some guesses.

Ping @mikelikespie

**Comment by Quarezz** [comment]
Same for me. 
If socket get's closed via internal error, e.g. internet connection - then there is no way of turning it back on. As state of socket becomes SR_CLOSED. 
And then - I call *connect* method that calls *open* and everything crashes as _readyState is still in SR_CLOSED after socket went down

**Comment by Quarezz** [comment]
So don't know if it's a SocketRocket issue or yours...

**Comment by accfews** [comment]
Yes I feel design smells about this issue, there is an incremental order with states and there is a check you can't set a smaller value. I don't know why its implemented that way. If there is no special reason for implementing this way I can fix it. But since there is a line like 

@"Cannot call -(void)open on SRWebSocket more than once");

Best solution for now is 
if (socket.readyState > SR_OPEN)
{
// re init the socket
}

**Comment by Quarezz** [comment]
Yup, will stick with reiniting the whole thing for now.

**Comment by SeraZheng** [comment]
I have got similar question.When socket was opening, program received stream event "NSStreamEventEndEncountered". And this property "_readyState" was assigned to "SR_CLOSED" within the delegate method "- (void)stream:(NSStream *)aStream handleEvent:(NSStreamEvent)eventCode" as follow:

case NSStreamEventEndEncountered: {
                [self _pumpScanner];
                SRFastLog(@"NSStreamEventEndEncountered %@", aStream);
                if (aStream.streamError) {
                    [self _failWithError:aStream.streamError];
                } else {
                    if (self.readyState != SR_CLOSED) {
                        self.readyState = SR_CLOSED;
                        _selfRetain = nil;
                    }

                    if (!_sentClose && !_failed) {
                        _sentClose = YES;
                        // If we get closed in this state it's probably not clean because we should be sending this when we send messages
                        dispatch_async(_callbackQueue, ^{
                            if ([self.delegate respondsToSelector:@selector(webSocket:didCloseWithCode:reason:wasClean:)]) {
                                [self.delegate webSocket:self didCloseWithCode:0 reason:@"Stream end encountered" wasClean:NO];
                            }
                        });
                    }
                }
                
                break;
            }

And then - I call open and everything crashes as _readyState is still in SR_CLOSED after socket went down. However, I would not like to init one instance of SRWebSocket again within third-party source code.

**Comment by wzbozon** [comment]
I solved a problem with opens another way that is present in this open-source examples: https://github.com/elabs/mobile-websocket-example
